k8srm-prototype/pkg/api/capacity_types_alternate.go

```diff
@@ -0,0 +1,177 @@
+type DevicePool struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   DevicePoolSpec   `json:"spec,omitempty"`
+	Status DevicePoolStatus `json:"status,omitempty"`
+}
+
+// DevicePoolSpec identifies the driver and contains the data for the pool
+// prior to any allocations.
+// NOTE: It's not clear that spec/status is the right model for this data.
+type DevicePoolSpec struct {
+	// NodeName is the name of the node containing the devices in the pool.
+	// For network attached devices, this may be empty.
+	// +optional
+	NodeName *string `json:"nodeName,omitempty"`
+
+	// Driver is the name of the DeviceDriver that created this object and
+	// owns the data in it.
+	//
+	// +required
+	DriverName string `json:"driverName"`
+
+	// Attributes contains device attributes that are common to all devices
+	// in the pool.
+	//
+	// +optional
+	Attributes []Attribute `json:"attributes,omitempty"`
+
+	// SharedResources are resources that are shared by all devices in the
+	// pool. This is typically used when representing a partitionable
+	// device, and need not be populated otherwise.
+	//
+	// +optional
+	SharedResources []DeviceResource `json:"sharedResources,omitempty"`
+
+	// Devices contains the individual devices in the pool.
+	//
+	// +required
+	Devices []Device `json:"devices"`
+}
+
+// DevicePoolStatus contains the state of the pool as last reported by the
+// driver. Note that this will not include the allocations that have been made
+// by the scheduler but not yet seen by the driver. Thus, it is NOT sufficient
+// to make future scheduling decisions.
+type DevicePoolStatus struct {
+	AllocatedDevices []AllocatedDevice `json:"allocatedDevices,omitempty"`
+}
+
+// AllocatedDevice represents a device that has been allocated from the pool.
+type AllocatedDevice struct {
+	Name string
+	ClaimUID types.UID
+}
+
+// Device is used to track individual devices in a pool.
+type Device struct {
+	// Name is a driver-specific identifier for the device.
+	//
+	// +required
+	Name string `json:"name"`
+
+	// Attributes contain additional metadata that can be used in
+	// constraints. If an attribute name overlaps with the pool attribute,
+	// the device attribute takes precedence.
+	//
+	// +optional
+	Attributes []Attribute `json:"attributes,omitempty"`
+
+	// SharedResourceRequests contains requests for some subset of the shared
+	// resources available in the overall pool. They represent the set of
+	// shared resources consumed by this device when it gets allocated.
+	//
+	// +optional
+	SharedResourceRequests []DeviceResourceRequest `json:"sharedResourceRequests,omitempty"`
+}
+
+// DeviceResource represents a named resource that is consumable by a device.
+// We indirect through DeviceResourceValue to enforce a one-of-many for the
+// different types of resources this struct can represent.
+type DeviceResource struct {
+	Name string `json:"name"`
+	*DeviceResourceValue `json:",inline"`
```

Comment:
I've noticed that this indirection (and the others like it) still haven't made it into the prototype or been discussed much. I really feel like limiting ourselves to *just* a quantity is going to bite us in the future.

Reply:
I agree, this is probably needed. We'll need to convince @thockin that the complexity is worth it. I think that given that the "capacity" model is primarily for driver developers, a little more complexity is OK. Users will have to be able to understand it though for troubleshooting.

Can you make sure this covered in the "open questions" doc?

Reply:
I'm not sure I understand what this is?